### PR TITLE
MueLu: half precision preconditioner

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/muelu 

## Motivation
Allow construction of a double-float MueLu preconditioner by setting "half precision" = true in the Stratimikos input list.

### CODE REVIEW

**Comment by cgcgcg** [comment]
FYI @jennloe

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_decl.hpp [hunk 1]
```diff
@@ -95,8 +96,8 @@
 
 namespace Thyra {
 
-  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
-  bool replaceWithXpetra(ParameterList& paramList, std::string parameterName);
+  // template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  // bool replaceWithXpetra(ParameterList& paramList, std::string parameterName);
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_decl.hpp:100]: why not just delete these?

File: packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp [hunk 2]
```diff
@@ -80,10 +81,12 @@
 #include <MueLu_ParameterListInterpreter.hpp>
 #include <MueLu_MLParameterListInterpreter.hpp>
 #include <MueLu_MasterList.hpp>
-#include <MueLu_XpetraOperator_decl.hpp> // todo fix me
+// #include <MueLu_XpetraOperator_decl.hpp> // todo fix me
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp:84]: why not delete this?

File: packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp [hunk 3]
```diff
@@ -59,171 +59,171 @@ namespace Thyra {
...
-        return true;
-      }
-      else {
-        TEUCHOS_TEST_FOR_EXCEPTION(true, MueLu::Exceptions::RuntimeError, "Parameter " << parameterName << " has wrong type.");
-        return false;
-      }
-    } else
-      return false;
-  }
+  // template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuRefMaxwellPreconditionerFactory_def.hpp:62]: Why not delete all this?

**Review by jhux2 — COMMENTED** [review]
I see where an Epetra double matrix is converted to float.  Where's the matching Tpetra code?  I think I've just missed it somewhere.

File: packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp [hunk 4]
```diff
@@ -324,158 +330,144 @@ namespace Thyra {
...
+
+            // build a new MueLu RefMaxwell preconditioner
+            RCP<MueLu::Hierarchy<Scalar,LocalOrdinal,GlobalOrdinal,Node> > H = MueLu::CreateXpetraPreconditioner(A, paramList);
+            xpPrecOp = rcp(new MueLuXpOp(H));
+          }
 
-        H->SetupRe();
-      }
+      } else {
+        // reuse old MueLu hierarchy stored in MueLu Tpetra/Epetra operator and put in new matrix
```
Comment (jhux2) [inline packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp:409]: This comment is misleading ... should be only Tpetra, correct?

Comment (cgcgcg) [reply packages/muelu/adapters/stratimikos/Thyra_MueLuPreconditionerFactory_decl.hpp:409]: No, I think that's correct at that point. We do support reuse for Epetra. The check for half precision happens below.

**Review by jhux2 — APPROVED** [review]
lgtm

**Review by jhux2 — APPROVED** [review]
lgtm

**Pull request merged**